File: 8_Kyu_Kata/count-of-positives-slash-sum-of-negatives/JavaScript/countOfPositivesSlashSumOfNegatives.Test.js
// [object Object] - Count of positives / sum of negatives  [ ID: 576bb71bbbcf0951d5000044  (count-of-positives-slash-sum-of-negatives) ]
// URL: https://www.codewars.com/kata/576bb71bbbcf0951d5000044
// Category: REFERENCE  |  Tags: FUNDAMENTALS | ARRAYS | LISTS
// *****************************************************************************
const chai = require("chai");
const assert = chai.assert;
chai.config.truncateThreshold=0;

describe("Example tests", () => {
  it("Testing for fixed test 1", () => {
    let testData = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, -11, -12, -13, -14, -15];
    let actual = countPositivesSumNegatives(testData);
    let expected = [10, -65];
    assert.deepEqual(actual, expected);
  })
    
  it("Testing for fixed test 2", () => {
    let testData = [0, 2, 3, 0, 5, 6, 7, 8, 9, 10, -11, -12, -13, -14];
    let actual = countPositivesSumNegatives(testData);
    let expected = [8, -50];    
    assert.deepEqual(actual, expected);
  });

  it("Testing for wrong test 1", () => {
    let testData = null;
    let actual = countPositivesSumNegatives(testData);
    assert.deepEqual(actual, []);
  })
   
  it("Testing for wrong test 2", () => {
    let testData = [];
    let actual = countPositivesSumNegatives(testData);    
    assert.deepEqual(actual, []);
  });
});

describe("Random Tests", () => {
   function solution(input) {
        let result = [];

        if (input == null || input.length == 0)
        {
            return result;
        }
        
        result.push(0);
        result.push(0);
        
        for(let i = 0; i < input.length; i++)
        {
            if (input[i] > 0)
            {
                result[0]++;
            }
            else
            {
                result[1] += input[i];
            }
        }
        
        return result;
    }
  
    it("Testing for 100 random tests", () => {
      for (let i = 0; i < 100; i++)
      {
          let testData = [];

          let testDataSize = Math.floor(Math.random()*100);

          for (let y = 0; y < testDataSize; y++)
          {
              testData.push(Math.floor(Math.random() * 100) * (Math.random() < 0.5 ? 1 : -1));
          }

          let expected = solution(testData);
          let actual = countPositivesSumNegatives(testData);        
          assert.deepEqual(actual, expected, `Testing for input = ${JSON.stringify(testData)}`);
      }
    });
});

File: 8_Kyu_Kata/count-of-positives-slash-sum-of-negatives/Python/count_of_positives_slash_sum_of_negatives_test.py
# [object Object] - Count of positives / sum of negatives  [ ID: 576bb71bbbcf0951d5000044  (count-of-positives-slash-sum-of-negatives) ]
# URL: https://www.codewars.com/kata/576bb71bbbcf0951d5000044
# Category: REFERENCE  |  Tags: FUNDAMENTALS | ARRAYS | LISTS
# ******************************************************************************
import codewars_test as test
from solution import count_positives_sum_negatives

@test.describe("Fixed Tests")
def basic_tests():
    @test.it('Basic Test Cases')
    def basic_test_cases():
        test.assert_equals(count_positives_sum_negatives([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, -11, -12, -13, -14, -15]),[10,-65])
        test.assert_equals(count_positives_sum_negatives([0, 2, 3, 0, 5, 6, 7, 8, 9, 10, -11, -12, -13, -14]),[8,-50])
        test.assert_equals(count_positives_sum_negatives([1]),[1,0])
        test.assert_equals(count_positives_sum_negatives([-1]),[0,-1])
        test.assert_equals(count_positives_sum_negatives([0,0,0,0,0,0,0,0,0]),[0,0])
        test.assert_equals(count_positives_sum_negatives([]),[])

@test.describe("Random Tests")
def random_tests():
    
    from random import randint
    from functools import reduce
    
    sol=lambda arr: reduce(lambda a,b: [a[0]+1 if b>0 else a[0],a[1]+b if b<0 else a[1]], arr, [0,0])

    for _ in range(40):
        arr=[randint(-100,100) for q in range(randint(1,100))]
        @test.it(f"Testing for count_positives_sum_negatives({arr})")
        def test_case():
            expected = sol(arr)
            test.assert_equals(count_positives_sum_negatives(arr), expected,"It should work for random inputs too")
